runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/AbstractWindmillStream.java

```diff
@@ -49,46 +44,64 @@
  * However, some accessors used for status page and other debugging need to take care not to require
  * synchronizing on this.
+ *
+ * <p>{@link #start()} and {@link #shutdown()} are called once in the lifetime of the stream. Once
+ * {@link #shutdown()}, a stream in considered invalid and cannot be restarted/reused.
  */
 public abstract class AbstractWindmillStream<RequestT, ResponseT> implements WindmillStream {
 
-  public static final long DEFAULT_STREAM_RPC_DEADLINE_SECONDS = 300;
   // Default gRPC streams to 2MB chunks, which has shown to be a large enough chunk size to reduce
   // per-chunk overhead, and small enough that we can still perform granular flow-control.
   protected static final int RPC_STREAM_CHUNK_SIZE = 2 << 20;
-  private static final Logger LOG = LoggerFactory.getLogger(AbstractWindmillStream.class);
-  protected final AtomicBoolean clientClosed;
-  private final AtomicBoolean isShutdown;
-  private final AtomicLong lastSendTimeMs;
-  private final Executor executor;
+  // Indicates that the logical stream has been half-closed and is waiting for clean server
+  // shutdown.
+  private static final Status OK_STATUS = Status.fromCode(Status.Code.OK);
+  protected final Sleeper sleeper;
+
+  /**
+   * Used to guard {@link #start()} and {@link #shutdown()} behavior.
+   *
+   * @implNote Do not hold when performing IO. If also locking on {@code this} in the same context,
+   *     should acquire shutdownLock first to prevent deadlocks.
+   */
+  protected final Object shutdownLock = new Object();
+
+  private final Logger logger;
+  private final ExecutorService executor;
   private final BackOff backoff;
-  private final AtomicLong startTimeMs;
-  private final AtomicLong lastResponseTimeMs;
-  private final AtomicInteger errorCount;
-  private final AtomicReference<String> lastError;
-  private final AtomicReference<DateTime> lastErrorTime;
-  private final AtomicLong sleepUntil;
   private final CountDownLatch finishLatch;
   private final Set<AbstractWindmillStream<?, ?>> streamRegistry;
   private final int logEveryNStreamFailures;
-  private final Supplier<StreamObserver<RequestT>> requestObserverSupplier;
-  // Indicates if the current stream in requestObserver is closed by calling close() method
-  private final AtomicBoolean streamClosed;
   private final String backendWorkerToken;
-  private @Nullable StreamObserver<RequestT> requestObserver;
+  private final ResettableStreamObserver<RequestT> requestObserver;
+  private final StreamDebugMetrics debugMetrics;
+  protected volatile boolean clientClosed;
+
+  /**
+   * Indicates if the current {@link ResettableStreamObserver} was closed by calling {@link
+   * #halfClose()}. Separate from {@link #clientClosed} as this is specific to the requestObserver
+   * and is initially false on retry.
+   */
+  @GuardedBy("this")
+  private boolean streamClosed;
+
+  private volatile boolean isShutdown;
+  private volatile boolean started;
```

Comment:
can this be guarded by one of hte locks instead of volatile?

Reply:
done